addons/io/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/internal/ModbusManagerImpl.java

```diff
@@ -404,22 +414,20 @@ private void returnConnection(ModbusSlaveEndpoint endpoint, Optional<ModbusSlave
         ModbusSlaveEndpoint endpoint = task.getEndpoint();
 
         ModbusRequestBlueprint request = task.getRequest();
-        Optional<ModbusSlaveConnection> connection = borrowConnection(endpoint);
+        Optional<ModbusSlaveConnection> connection = timer.connection.timeSupplier(() -> borrowConnection(endpoint));
         logger.trace("Executing task {} (oneOff={})! Connection received in {} ms [operation ID {}]", task, oneOffTask,
                 System.currentTimeMillis() - connectionBorrowStart, operationId);
-        ExecutorService callbackPool = scheduledThreadPoolExecutor;
-        if (callbackPool == null) {
+        if (scheduledThreadPoolExecutor == null) {
             // manager deactivated
-            invalidate(endpoint, connection);
+            timer.connection.timeRunnable(() -> invalidate(endpoint, connection));
             return Optional.empty();
         }
         if (!connection.isPresent()) {
             logger.warn("Could not connect to endpoint {} -- aborting request {} [operation ID {}]", endpoint, request,
                     operationId);
             if (callback != null) {
-                callbackPool.execute(() -> {
```

Comment:
We want to run in some thread right away, similar to other callbacks

Should have spotted this already in https://github.com/openhab/openhab2-addons/pull/4158